runtime/bcutil/cfreader.c

```diff
@@ -2195,18 +2195,35 @@ checkClass(J9PortLibrary *portLib, J9CfrClassFile* classfile, U_8* segment, U_32
 		goto _errorFound;
 	}
 
+	/* Java 9 spec: If the ACC_INTERFACE flag is not set, ACC_MODULE may not be set. */
+	if ((flags & BCT_MajorClassFileVersionMask) >= BCT_Java9MajorVersionShifted) {
+		if ((value & CFR_ACC_MODULE) && (! (value & CFR_ACC_INTERFACE))) {
+			errorCode = J9NLS_CFR_ERR_MODULE_NOT_INTERFACE__ID;
+			offset = endOfConstantPool;
+			goto _errorFound;
+		}
+	}
+
 	if((value & CFR_ACC_FINAL)&&(value & CFR_ACC_ABSTRACT)) {
 		errorCode = J9NLS_CFR_ERR_FINAL_ABSTRACT_CLASS__ID;
 		offset = endOfConstantPool;
 		goto _errorFound;
 	}
 
-	/* check annotations and interfaces for invalid flags */
-	if ((value & CFR_ACC_INTERFACE) && (value & ~CFR_INTERFACE_CLASS_ACCESS_MASK))
+	/* check annotations and interfaces for invalid flags. If the ACC_INTERFACE flag is set,
+	 * the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.
+	 *
+	 * For Java 9 ACC_MODULE also must not be set
+	 */
+	 if (J9_ARE_ALL_BITS_SET(value, CFR_ACC_INTERFACE))
 	{
-		errorCode = J9NLS_CFR_ERR_INTERFACE_CLASS__ID;
-		offset = endOfConstantPool;
-		goto _errorFound;
+		 if ((((flags & BCT_MajorClassFileVersionMask) >= BCT_Java9MajorVersionShifted)
+				 && J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK_9))
+			|| J9_ARE_ANY_BITS_SET(value, CFR_INTERFACE_FORBIDDEN_ACCESS_MASK)) {
```

Comment:
Please move the `) {` to the next line for this multi-line if statement. It should line up with the if.